src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,257 @@
+
+    private readonly DualModeKalmanFilter<float> _cpuUsageFilter = new();
+    private readonly DualModeKalmanFilter<float> _availableMemoryFilter = new();
+    private readonly DualModeKalmanFilter<long> _memoryUsageFilter = new();
+
+    private Task<SiloAddress> _cachedLocalSilo;
+
+    public ResourceOptimizedPlacementDirector(
+        DeploymentLoadPublisher deploymentLoadPublisher,
+        IOptions<ResourceOptimizedPlacementOptions> options)
+    {
+        _options = options.Value;
+        deploymentLoadPublisher?.SubscribeToStatisticsChangeEvents(this);
+    }
+
+    public Task<SiloAddress> OnAddActivation(PlacementStrategy strategy, PlacementTarget target, IPlacementContext context)
+    {
+        var compatibleSilos = context.GetCompatibleSilos(target);
+
+        if (IPlacementDirector.GetPlacementHint(target.RequestContextData, compatibleSilos) is { } placementHint)
+        {
+            return Task.FromResult(placementHint);
+        }
+
+        if (compatibleSilos.Length == 0)
+        {
+            throw new SiloUnavailableException($"Cannot place grain with Id = [{target.GrainIdentity}], because there are no compatible silos.");
+        }
+
+        if (compatibleSilos.Length == 1)
+        {
+            return Task.FromResult(compatibleSilos[0]);
+        }
+
+        if (siloStatistics.IsEmpty)
+        {
+            return Task.FromResult(compatibleSilos[Random.Shared.Next(compatibleSilos.Length)]);
+        }
+
+        var bestCandidate = GetBestSiloCandidate(compatibleSilos);
+
+        if (IsLocalSiloPreferable(context, compatibleSilos, bestCandidate.Value))
+        {
+            return _cachedLocalSilo ??= Task.FromResult(context.LocalSilo);
+        }
+
+        return Task.FromResult(bestCandidate.Key);
+    }
+
+    private KeyValuePair<SiloAddress, float> GetBestSiloCandidate(SiloAddress[] compatibleSilos)
+    {
+        List<KeyValuePair<SiloAddress, ResourceStatistics>> relevantSilos = [];
+        foreach (var silo in compatibleSilos)
+        {
+            if (siloStatistics.TryGetValue(silo, out var stats) && !stats.IsOverloaded)
+            {
+                relevantSilos.Add(new(silo, stats));
+            }
+        }
+
+        int chooseFrom = (int)Math.Ceiling(Math.Sqrt(relevantSilos.Count));
+        Dictionary<SiloAddress, float> chooseFromSilos = [];
+
+        while (chooseFromSilos.Count < chooseFrom)
+        {
+            int index = Random.Shared.Next(relevantSilos.Count);
+            var pickedSilo = relevantSilos[index];
+
+            relevantSilos.RemoveAt(index);
+
+            float score = CalculateScore(pickedSilo.Value);
+            chooseFromSilos.Add(pickedSilo.Key, score);
+        }
+
+        var orderedByLowestScore = chooseFromSilos.OrderBy(kv => kv.Value);
+
+        // there could be more than 1 silo that has the same score, we pick 1 of them randomly so that we dont continuously pick the first one.
```

Comment:
How likely is this given the values are floating point?

Reply:
Adding a bit of random jitter to the scores as they are calculated is an alternative approach which should help us to avoid most of this work

Reply:
> How likely is this given the values are floating point?

Not very likely, but can happen